### PR TITLE
Adding filter and search functionality for messages list API (closes #4)

### LINKED ISSUE #4: Message search functionality
Add the ability to search for messages within a room based on keywords, user, or date range.

### PR DESCRIPTION
**Pull request by pesfahanian**
Added all the requirements in the issue description.
Code style is kept as same and as untouched as possible.
Every new thing has tests and docstrings/comments.
All tests pass.

### CODE REVIEW

**Comment by pesfahanian** [comment]
Added all the requirements in the issue description.
Code style is kept as close and as untouched as possible.
Every new thing has tests.
All tests pass.

File: df_chat/drf/serializers.py [hunk 1]
```diff
@@ -249,5 +249,5 @@ class UserNameSerializer(serializers.ModelSerializer):
 
     class Meta:
         model = User
-        read_only_fields = ("display_name",)
```
Comment (pesfahanian) [inline df_chat/drf/serializers.py:252]: This was breaking the Swagger UI.

Comment (sreekanthreddybalne) [reply df_chat/drf/serializers.py:252]: Right. There is no "display_name" on the User model. Could you please replace it with "username"?

File: changes.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# Changes
+
+- Added `drf-yasg` for swagger API view.
+- Fixed `read_only_fields` for `UserNameSerializer` serializer.
+- Added `django-filters` for `Message` filter backend.
+- Added tests for `Message` list API.
```
Comment (sreekanthreddybalne) [inline changes.md:6]: Could you please explain why this file is needed and provide an explanation in the commit description?

File: df_chat/tests/test_chat.py [hunk 3]
```diff
@@ -247,6 +249,88 @@ async def test_chat_multiple_rooms(self):
         await communicator2.disconnect()
         await communicator3.disconnect()
 
+    def test_create_message(self):
+        User = get_user_model()
```
Comment (sreekanthreddybalne) [inline df_chat/tests/test_chat.py:253]: Could you please add a doc string that explains the steps required to create a message?

File: df_chat/tests/test_chat.py [hunk 4]
```diff
@@ -247,6 +249,88 @@ async def test_chat_multiple_rooms(self):
...
+        user = User.objects.create_superuser(username="test_username", password="test_password")
```
Comment (sreekanthreddybalne) [inline df_chat/tests/test_chat.py:254]: Please reuse the "TestChat.create_user" method, to create a user.

File: df_chat/tests/test_chat.py [hunk 5]
```diff
@@ -247,6 +249,88 @@ async def test_chat_multiple_rooms(self):
...
+        )
+        self.assertEqual(room_create_response.status_code, http.HTTPStatus.CREATED)
+        room_id = room_create_response.json()["id"]
+        message_create_response = self.client.post(
+            f"/api/v1/chat/rooms/{room_id}/messages/",
+            data={"body": "Test message 1234."},
+        )
+        self.assertEqual(message_create_response.status_code, http.HTTPStatus.CREATED)
+    
+    def test_list_message(self):
```
Comment (sreekanthreddybalne) [inline df_chat/tests/test_chat.py:268]: Could you please add a docstring that explains the purpose of the test you are running here?

File: df_chat/tests/test_chat.py [hunk 6]
```diff
@@ -247,6 +249,88 @@ async def test_chat_multiple_rooms(self):
...
+        self.assertEqual(messages_list_response.status_code, http.HTTPStatus.OK)
+        self.assertEqual(messages_list_response.json(), [])
+
+        # * date range search
+        messages_list_response = self.client.get(
+                f"/api/v1/chat/rooms/{room_id}/messages/",
+                {"created_lte": str(datetime.datetime.now())},
+            )
+        self.assertEqual(messages_list_response.status_code, http.HTTPStatus.OK)
+        self.assertNotEqual(messages_list_response.json(), [])
```
Comment (sreekanthreddybalne) [inline df_chat/tests/test_chat.py:326]: Rather than testing that the returned JSON is not empty, we should verify that the data returned was created on or before the date specified in the filter. 

The same applies to the 'username', 'body', and 'created_gte' filters. To accomplish this, please create a few sample messages and apply the filters to them.

**Review by sreekanthreddybalne — CHANGES_REQUESTED** [review]
Few points to note while creating a pull request:
- Please ensure that the pull request title accurately describes the issue being addressed. For example, a concise title like 'Implement filter and search functionality for messages' is more appropriate than simply stating 'Closes #4'. The title should be attention-grabbing enough so that the reviewer does not need to refer back to the issue. 
  However, it would be helpful to include a link to the issue in the pull request description.
- Please note that developers often refer to commit descriptions to understand why something was added or removed. Therefore, it is essential to include the purpose or rationale behind the changes in the commit description.

By the way, great job! Please concentrate on improving the commenting in the code and taking into account edge cases when writing unit tests.

File: df_chat/drf/serializers.py (hunk 1 above)
Comment (pesfahanian) [reply df_chat/drf/serializers.py:252]: Done!

File: changes.md (hunk 2 above)
Comment (pesfahanian) [reply changes.md]: Was there for the reviewer's convenience. Removed!

File: df_chat/tests/test_chat.py (hunk 3 above)
Comment (pesfahanian) [reply df_chat/tests/test_chat.py]: Done!

File: df_chat/tests/test_chat.py (hunk 4 above)
Comment (pesfahanian) [reply df_chat/tests/test_chat.py]: Done!

File: df_chat/tests/test_chat.py (hunk 5 above)
Comment (pesfahanian) [reply df_chat/tests/test_chat.py]: Done!

File: df_chat/tests/test_chat.py (hunk 6 above)
Comment (pesfahanian) [reply df_chat/tests/test_chat.py]: Done!

File: df_chat/tests/test_chat.py [hunk 7]
```diff
@@ -6,19 +6,27 @@
...
-from django.contrib.auth import get_user_model
+from datetime import datetime, timedelta
 import http
-import datetime
+from typing import Tuple
+import os
 
 class TestChat(TransactionTestCase):
     """
-    Test for chat appplication
```
Comment (pesfahanian) [inline df_chat/tests/test_chat.py:17]: There was a spelling mistake here.

File: df_chat/drf/filters.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+from ..models import Message
+
+class MessageFilter(django_filters.rest_framework.FilterSet):
+    """
+    Filter-set class for `Message` list API.
+    All the filters that will be accessible by the filter backend
+        are defined here.
+    """
+    # For `datetime` range filter on the model's `created` field.
+    created_gte = django_filters.rest_framework.DateTimeFilter(
```
Comment (eugapx) [inline df_chat/drf/filters.py:12]: Please rewrite imports as it suggested in [docs](https://django-filter.readthedocs.io/en/stable/guide/rest_framework.html)  

```
from django_filters import rest_framework as filters
```

File: df_chat/tests/test_chat.py [hunk 9]
```diff
@@ -2,16 +2,48 @@
...
+from df_chat.models import User
+from df_chat.tests.utils import RoomFactory
+from df_chat.tests.utils import TEST_USER_PASSWORD
+from df_chat.tests.utils import UserFactory
+from django.test import TransactionTestCase
+from rest_framework.reverse import reverse
+from tests.asgi import application
+from typing import Tuple
 from df_chat.tests.base import BaseTestUtilsMixin
 from django.test import TransactionTestCase
```
Comment (pesfahanian) [inline df_chat/tests/test_chat.py:14]: This import already exists on line 9.

**Comment by pesfahanian** [comment]
Recent changes to the tests break the following tests:
- `TestChat.test_auth`
- `TestChat.test_chat_multiple_rooms`
- `TestChat.test_chat_single_room`

The problem is with the new `BaseTestUtilsMixin.async_create_user` method.

Awaiting further instructions to fix the tests.

**Review by eugapx — CHANGES_REQUESTED** [review]
Please make sure to install pre-commit hooks: https://pre-commit.com/

File: df_chat/drf/filters.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        are defined here.
+    """
+    # For `datetime` range filter on the model's `created` field.
+    created_gte = filters.DateTimeFilter(
+        field_name="created", lookup_expr="gte")
+    created_lte = filters.DateTimeFilter(
+        field_name="created", lookup_expr="lte")
+    
+    # For filter on the model's related user's `username` field.
+    # TODO: Maybe a better/cleaner implementation for this?
```
Comment (eugapx) [inline df_chat/drf/filters.py:18]: Seems good to me. Do you have some specific concern on this code?

File: df_chat/drf/filters.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    username = filters.CharFilter(
+        field_name="room_user__user__username", lookup_expr="contains")
+    
+    # For text search on the model's `body` field.
+    # ! This is not efficient. Text search is better handled otherwise.
```
Comment (eugapx) [inline df_chat/drf/filters.py:23]: I don't think we will have some advanced text search in this package for now. 

If we will, I think, it's better to create an issue rather than adding a comment.

File: df_chat/models.py [hunk 12]
```diff
@@ -89,13 +89,28 @@ def get_upload_to(self, filename):
     image = models.ImageField(upload_to=get_upload_to, null=True, blank=True)
 
     is_public = models.BooleanField(default=True)
+    
+    # ! This is unnecessary when there exists a through-model `RoomUser` model.
```
Comment (eugapx) [inline df_chat/models.py:93]: Make sense. Please raise an issue instead

File: df_chat/models.py [hunk 13]
```diff
@@ -89,13 +89,28 @@ def get_upload_to(self, filename):
...
     users = models.ManyToManyField(User, blank=True)
+    
     admins = models.ManyToManyField(User, blank=True, related_name="rooms_admin_set")
 
     muted_by = models.ManyToManyField(User, blank=True, related_name="room_muted_set")
 
     objects = RoomQuerySet.as_manager()
 
+    @property
+    def users_count(self) -> int:
```
Comment (eugapx) [inline df_chat/models.py:103]: Why do we need this method? 

`room.users.count()` seems self-explanatory to me

File: requirements.txt [hunk 14]
```diff
@@ -23,3 +23,4 @@ django-df-auth
 factory-boy
 pytest
 pytest-django
+drf-yasg
```
Comment (eugapx) [inline requirements.txt:26]: We use drf-spectacular instead

File: df_chat/drf/filters.py [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+
+from ..models import Message
+
+class MessageFilter(filters.FilterSet):
+    """
+    Filter-set class for `Message` list API.
+    All the filters that will be accessible by the filter backend
+        are defined here.
+    """
+    # For `datetime` range filter on the model's `created` field.
```
[the remark this answers is not in the document]
Comment (pesfahanian) [reply df_chat/drf/filters.py]: Done!

File: df_chat/models.py (hunk 12 above)
Comment (pesfahanian) [reply df_chat/models.py]: Done! #21

File: df_chat/models.py (hunk 13 above)
Comment (pesfahanian) [reply df_chat/models.py]: Done!

File: requirements.txt (hunk 14 above)
Comment (pesfahanian) [reply requirements.txt:26]: Done!

File: tests/urls.py [hunk 16]
```diff
@@ -4,9 +4,30 @@
 from django.urls import include
 from django.urls import path
 
+from drf_yasg import openapi
```
[the remark this answers is not in the document]
Comment (pesfahanian) [reply tests/urls.py]: Done!
Added its configurations to `settings.py` and its routes to `urls.py`.

**Comment by pesfahanian** [comment]
> Please make sure to install pre-commit hooks: https://pre-commit.com/

Done! Also added its requirement to `requirements.txt`.

File: df_chat/drf/filters.py [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+    """
+
+    created_gte = filters.DateTimeFilter(field_name="created", lookup_expr="gte")
+    created_lte = filters.DateTimeFilter(field_name="created", lookup_expr="lte")
+
+    username = filters.CharFilter(
+        field_name="room_user__user__username", lookup_expr="contains"
+    )
+
+    body = filters.CharFilter(field_name="body", lookup_expr="contains")
```
Comment (sreekanthreddybalne) [inline df_chat/drf/filters.py:17]: How about using "icontains" here, to enable case insensitive search?

File: df_chat/tests/test_message_endpoint.py [hunk 18]
```diff
@@ -26,6 +31,110 @@ def test_message_creation_success(self):
...
+            "message",
+            "1234",
+        ]:
+            messages_list_response = await self.async_client.get(
+                messages_list_api_url,
+                {"body": body},
+            )
+            # Assert the correct HTTP status code for response.
+            self.assertEqual(messages_list_response.status_code, http.HTTPStatus.OK)
+            # Assert the response body contains the keyword.
```
Comment (sreekanthreddybalne) [inline df_chat/tests/test_message_endpoint.py:69]: Add a comment saying that the search is case insensitive here.

File: df_chat/tests/test_message_endpoint.py [hunk 19]
```diff
@@ -26,6 +31,110 @@ def test_message_creation_success(self):
...
+            "somethingrandom",
+            "messages",
+            "124",
+        ]:
+            messages_list_response = await self.async_client.get(
+                messages_list_api_url,
+                {"body": body},
+            )
+            # Assert the correct HTTP status code for response.
+            self.assertEqual(messages_list_response.status_code, http.HTTPStatus.OK)
```
Comment (sreekanthreddybalne) [inline df_chat/tests/test_message_endpoint.py:85]: Please asset that the messages_list_response.json() is empty list, as these keywords are not present in any of the messages.

File: df_chat/tests/test_message_endpoint.py [hunk 20]
```diff
@@ -26,6 +31,110 @@ def test_message_creation_success(self):
...
+            messages_list_response.json()[0]["room_user_id"], room_user_obj.id
+        )
+
+        # * 2. username search - not ok case
+        messages_list_response = await self.async_client.get(
+            messages_list_api_url,
+            {"username": "invalid_username"},
+        )
+        # Assert the correct HTTP status code for response.
+        self.assertEqual(messages_list_response.status_code, http.HTTPStatus.OK)
```
Comment (sreekanthreddybalne) [inline df_chat/tests/test_message_endpoint.py:107]: Same here, please assert that the messages_list_response.json() is an empty list.

File: df_chat/tests/test_message_endpoint.py [hunk 21]
```diff
@@ -26,6 +31,110 @@ def test_message_creation_success(self):
...
+            messages_list_api_url,
+            {"created_gte": str(before)},
+        )
+        message_created = datetime.strptime(
+            messages_list_response.json()[0]["created"], datetime_format
+        )
+        # Assert the correct HTTP status code for response.
+        self.assertEqual(messages_list_response.status_code, http.HTTPStatus.OK)
+        # Assert message `created` datetime is larger than `before` datetime.
+        self.assertTrue(message_created >= before)
```
Comment (sreekanthreddybalne) [inline df_chat/tests/test_message_endpoint.py:124]: While running this test, the messages are always created in the past few minutes/seconds. So, this test will always pass.
Please add a message to the database that appears to have been created over an hour ago and check that the message is not returned in the response.

File: df_chat/tests/test_message_endpoint.py [hunk 22]
```diff
@@ -26,6 +31,110 @@ def test_message_creation_success(self):
...
+            messages_list_api_url,
+            {"created_lte": str(now)},
+        )
+        message_created = datetime.strptime(
+            messages_list_response.json()[0]["created"], datetime_format
+        )
+        # Assert the correct HTTP status code for response.
+        self.assertEqual(messages_list_response.status_code, http.HTTPStatus.OK)
+        # Assert message `created` datetime is smaller than `now` datetime.
+        self.assertTrue(message_created <= now)
```
Comment (sreekanthreddybalne) [inline df_chat/tests/test_message_endpoint.py:137]: Please add a test where  the filter is "{"created_lte": str(before)}", to ensure that the endpoint does not return any messages that are created after "before" time.

Comment (pesfahanian) [reply df_chat/tests/test_message_endpoint.py]: Done!

File: df_chat/tests/test_message_endpoint.py (hunk 21 above)
Comment (pesfahanian) [reply df_chat/tests/test_message_endpoint.py]: I made sure of this. The `datetime` format of the messages are `%Y-%m-%dT%H:%M:%S.%f` and event though the message creation and the list API call are close to each other, the time difference in the `%f` portion of the `created` field asserts that the message is indeed created before the API call.

File: df_chat/tests/test_message_endpoint.py (hunk 20 above)
Comment (pesfahanian) [reply df_chat/tests/test_message_endpoint.py]: Done!

File: df_chat/tests/test_chat.py [hunk 23]
```diff
@@ -2,16 +2,46 @@
...
+        return user, token
+
+    @database_sync_to_async
+    def create_room_and_add_users(self, *users):
+        """
+        Creates a Room object and adds the users to it.
+        """
+        room = RoomFactory()
+        room.users.set(users)
+        return room
```
[the remark this answers is not in the document]
Comment (pesfahanian) [reply df_chat/tests/test_chat.py]: Fixed!

File: df_chat/drf/filters.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (pesfahanian) [reply df_chat/drf/filters.py:17]: Exactly the reason I didn't use the `icontains`. I was referencing keyword search in apps like WhatsApp where it is NOT case-sensitive.

File: df_chat/tests/test_message_endpoint.py [hunk 24]
```diff
@@ -17,15 +22,141 @@ def test_message_creation_success(self):
 
         message_endpoint = reverse("rooms-messages-list", kwargs={"room_pk": room.pk})
         self.client.credentials(HTTP_AUTHORIZATION="Bearer " + token)
-        response = self.client.post(message_endpoint, {"body": "Hi"})
+        body = "Test message 1234."
+        response = self.client.post(message_endpoint, {"body": body})
```
Comment (pesfahanian) [inline df_chat/tests/test_message_endpoint.py:26]: Changed the `test_message_creation_success` test body to accommodate the following filtering tests.

File: df_chat/drf/filters.py [hunk 25]
```diff
@@ -0,0 +1,21 @@
...
+
+    class Meta:
+        model = Message
+        fields = ()
```
Comment (sreekanthreddybalne) [inline df_chat/drf/filters.py:21]: Do we have to set the "fields" here to an empty tuple? Could you please check what happens if we remove it completely?

File: df_chat/tests/test_message_endpoint.py [hunk 26]
```diff
@@ -17,15 +22,141 @@ def test_message_creation_success(self):
...
+        return user, room
+
+    def test_filter_message_by_keyword(self):
+        """
+        Test-case for filtering messages by the `Message` model's `body` field.
+        """
+        _, room = self.test_message_creation_success()
+        messages_list_endpoint = reverse(
+            "rooms-messages-list", kwargs={"room_pk": room.pk}
+        )
```
Comment (sreekanthreddybalne) [inline df_chat/tests/test_message_endpoint.py:46]: Please declare it as an attribute on the TestMessageEndpoint class and use it in all test methods.

File: df_chat/tests/test_message_endpoint.py [hunk 27]
```diff
@@ -17,15 +22,141 @@ def test_message_creation_success(self):
...
+            "messages",
+            "124",
+        ]:
+            messages_list_response = self.client.get(
+                messages_list_endpoint,
+                {"body": body},
+            )
+            # Assert the correct HTTP status code for response.
+            self.assertEqual(messages_list_response.status_code, http.HTTPStatus.OK)
+            # Assert the response is an empty list.
```
Comment (sreekanthreddybalne) [inline df_chat/tests/test_message_endpoint.py:79]: Please remove these comments. It is implicit in the statement itself.

Also, please do the same in other parts of the tests.

File: df_chat/tests/test_message_endpoint.py [hunk 28]
```diff
@@ -17,15 +22,141 @@ def test_message_creation_success(self):
...
+        messages_list_response = self.client.get(
+            messages_list_endpoint,
+            {"username": "invalid_username"},
+        )
+        # Assert the correct HTTP status code for response.
+        self.assertEqual(messages_list_response.status_code, http.HTTPStatus.OK)
+        # Assert the response is an empty list.
+        self.assertEqual(messages_list_response.json(), [])
+
+    def test_filter_message_by_date_range(self):
```
Comment (sreekanthreddybalne) [inline df_chat/tests/test_message_endpoint.py:113]: Please test the edge cases of the filter. 
- The sample data should include messages that fall within a range as well as out of the range of the filter.
- Then in each test, make sure that the response only returns the data from within the requested range.

In this case, I see that only one message is used as a sample data.  Please increase the sample size to atleast 3 messages such that 2 of them fall within the requested range.

**Review by sreekanthreddybalne — CHANGES_REQUESTED** [review]
Hello @pesfahanian, thank you for your work on the PR. Could you please address all the comments before requesting a re-review? This will help ensure that the review process is as efficient as possible. Thank you for your understanding.

File: df_chat/tests/test_message_endpoint.py (hunk 28 above)
Comment (pesfahanian) [reply df_chat/tests/test_message_endpoint.py:95]: Done! The new scenarios in the `TestMessageEndpoint.test_filter_message_by_date_range` should cover all of these. It creates multiple messages in-between multiple timestamps and checks the API results against all of them.

Also, since messages created via to back-to-back calls to `TestMessageEndpoint.test_message_creation_success` do not persist (due to database rollback after each test-case), messages are created manually in  `TestMessageEndpoint.test_filter_message_by_date_range`.

An alternative is to write a custom test runner that can exclude single test cases from database rollback, but this needs approval.

File: df_chat/tests/test_message_endpoint.py (hunk 27 above)
Comment (pesfahanian) [reply df_chat/tests/test_message_endpoint.py]: Done, though a couple of more complicated assertions needed a comment in my opinion.

File: df_chat/tests/test_message_endpoint.py (hunk 26 above)
Comment (pesfahanian) [reply df_chat/tests/test_message_endpoint.py]: Done!

File: df_chat/drf/filters.py (hunk 25 above)
Comment (pesfahanian) [reply df_chat/drf/filters.py:21]: Removing it causes an error (Meta classes need to have a `fields` attribute declared).
Also added `icontains`.

**Comment by pesfahanian** [comment]
> Hello @pesfahanian, thank you for your work on the PR. Could you please address all the comments before requesting a re-review? This will help ensure that the review process is as efficient as possible. Thank you for your understanding.

All the comments have been addressed.
